pandas/core/algorithms.py

```diff
@@ -211,47 +211,47 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
         except TypeError:
             raise TypeError("bins argument only works with numeric data.")
         values = cat.codes
-    elif is_category:
-        bins = values.categories
-        cat = values
-        values = cat.codes
 
-    dtype = values.dtype
+    if is_category:
 
-    if issubclass(values.dtype.type, (np.datetime64, np.timedelta64)) or is_period:
-        if is_period:
-            values = PeriodIndex(values)
+        result = values.value_counts(dropna)
 
-        values = values.view(np.int64)
-        keys, counts = htable.value_count_int64(values)
+    else:
 
-        if dropna:
-            from pandas.tslib import iNaT
-            msk = keys != iNaT
-            keys, counts = keys[msk], counts[msk]
-        # convert the keys back to the dtype we came in
-        keys = keys.astype(dtype)
+        dtype = values.dtype
 
-    elif com.is_integer_dtype(dtype):
-        values = com._ensure_int64(values)
-        keys, counts = htable.value_count_int64(values)
+        if issubclass(values.dtype.type, (np.datetime64, np.timedelta64)) or is_period:
```

Comment:
you can use ``is_datetimelike_array`` (which may not pass all tests), or ``is_datetime_or_timedelta_dtype`` which is a bit more strict (you can define this like ``is_period`` is above)